cpp/src/strings/convert/convert_urls.cu

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/strings/convert/convert_urls.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace
+{
+
+//
+// This is the functor for the url_encode() method below.
+// Specific requirements are documented in custrings issue #321.
+// In summary it converts mostly non-ascii characters and control characters into UTF-8 hex characters
+// prefixed with '%'. For example, the space character must be converted to characters '%20' where the
+// '20' indicates the hex value for space in UTF-8. Likewise, multi-byte characters are converted to
+// multiple hex charactes. For example, the é character is converted to characters '%C3%A9' where 'C3A9'
+// is the UTF-8 bytes xc3a9 for this character.
+//
+struct url_encoder_fn
+{
+    column_device_view const d_strings;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    // utility to create 2-byte hex characters from single binary byte
+    __device__ void byte_to_hex( u_char byte, char* hex )
+    {
+        hex[0] = '0';
+        if( byte >= 16 )
+        {
+            u_char hibyte = byte/16;
+            hex[0] = hibyte < 10 ? '0'+hibyte : 'A'+(hibyte-10);
+            byte = byte - (hibyte * 16);
+        }
+        hex[1] = byte < 10 ? '0'+byte : 'A'+(byte-10);
+    }
+
+    // main part of the functor the performs the url-encoding
+    __device__ size_type operator()( size_type idx )
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        //
+        char* out_ptr = nullptr;
+        if( d_chars )
+            out_ptr = d_chars + d_offsets[idx];
```

Comment:
Change to: `char* out_ptr = d_chars ? out_ptr = d_chars + d_offsets[idx] ? nullptr;`